### PR TITLE
[FAB-32] Undo action

### PR DESCRIPTION
**Pull request by daystram**
- [x] Undo and Checkpoint WSMessageType backend handler
- [x] Paint and fill tool undo
- [x] Text tool undo 
- [x] Improved text tool UI
- [x] Improved text layer rendering performance 
- [x] Improved Hub clickable text TTS UI
- [x] Undo on touchscreen devices
  - onPointerUp() not triggered on touchscreen tap events

Known bugs:
- Undoing the first checkpoint requires another undo action, which only happens once every component mount
  - Cause: react's useCallback() memoisation + async setState() behaviour
- Undoing text edits leaves the first added/editing character in place, despite bounding box update properly
  - Cause: no clue m8. @nicolauscg tasukete

### CODE REVIEW

File: fableous-be/handlers/session.go [hunk 1]
```diff
@@ -25,9 +25,10 @@ func (m *module) SessionGetAllByClassroomID(classroomID string) (sessionInfos []
 			ID:          session.ID,
 			Title:       session.Title,
 			Description: session.Description,
-			Pages:       session.Pages,
-			Completed:   session.Completed,
-			CreatedAt:   session.CreatedAt,
+			// TODO
```
Comment (nicolauscg) [inline fableous-be/handlers/session.go:28]: is todo here for adding names of student?

Comment (daystram) [reply fableous-be/handlers/session.go]: Fixed

File: fableous-fe/src/components/canvas/Canvas.tsx [hunk 2]
```diff
@@ -261,15 +277,37 @@ const Canvas = forwardRef<HTMLCanvasElement, CanvasProps>(
...
-      Object.entries(textShapes).forEach(([id, shape]) => {
+      Object.entries(textShapesRef.current).forEach(([id, shape]) => {
         ctx.font = `${shape.fontSize * SCALE}px Arial`;
         ctx.fillText(
           shape.text,
           (shape.x1 + shape.x2) / 2,
           (shape.y1 + shape.y2) / 2
         );
-        if (parseInt(id, 10) === editingTextId) {
+        if (parseInt(id, 10) === editingTextIdRef.current) {
```
Comment (nicolauscg) [inline fableous-fe/src/components/canvas/Canvas.tsx:287]: what does the body of this conditional and the one with FRAME_COUNTER do?

**Review by nicolauscg — APPROVED** [review]
all good aside a clarification qs

File: fableous-fe/src/components/canvas/Canvas.tsx (hunk 2 above)
Comment (daystram) [reply fableous-fe/src/components/canvas/Canvas.tsx:287]: this highlights the texts in bright green boxes when a textShape is selected. the one with FRAME_COUNTER adds a blinking caret on controllers using ToolMode.Text -- just adding a signifier that they are to edit the selected text

**Pull request merged**